src/Analyzers/CSharp/Analyzers/SimplifyLinqExpressions/CSharpSimplifyLinqExpressionsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Operations;
+using System.Linq;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System.Linq.Expressions;
+#if CODE_STYLE
+using OptionSet = Microsoft.CodeAnalysis.Diagnostics.AnalyzerConfigOptions;
+#endif
+
+namespace Microsoft.CodeAnalysis.CSharp.SimplifyLinqExpressions
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpSimplifyLinqExpressionsDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        internal readonly ImmutableArray<string> s_validLinqCalls = ImmutableArray.Create(
+                nameof(Enumerable.First),
+                nameof(Enumerable.Last),
+                nameof(Enumerable.Single),
+                nameof(Enumerable.Any),
+                nameof(Enumerable.Count),
+                nameof(Enumerable.SingleOrDefault),
+                nameof(Enumerable.FirstOrDefault),
+                nameof(Enumerable.LastOrDefault)
+            );
+
+        public CSharpSimplifyLinqExpressionsDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.SimplifyLinqExpressionsDiagnosticId,
+                   option: null,
+                   title: new LocalizableResourceString(nameof(CSharpAnalyzersResources.Simplify_linq_expression), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var whereMethods = ImmutableHashSet.CreateBuilder<IMethodSymbol>();
+            var linqMethods = ImmutableHashSet.CreateBuilder<IMethodSymbol>();
+
+            var enumNamedType = context.Compilation.GetTypeByMetadataName(typeof(System.Linq.Enumerable).FullName);
+            if (enumNamedType is null)
+            {
+                return;
+            }
+
+            var enumMethods = enumNamedType.GetMembers(nameof(Enumerable.Where)).OfType<IMethodSymbol>();
+
+            if (enumMethods is null)
+            {
+                return;
+            }
+
+            AddIfNotNull(whereMethods, enumMethods);
+
+            // add all valid linq calls
+            foreach (var id in s_validLinqCalls)
+            {
+                enumMethods = enumNamedType?.GetMembers(id).OfType<IMethodSymbol>();
```

Comment:
enumNamedType can't be null.  

this also means enumMethods can't be null.  so you don't need the AddIfNotNull call.